### PR TITLE
Non prescriptive terms for standard config introduced

### PR DESCRIPTION
**Pull request by blmalone**
**Description**

As per @mds1's pull request: https://github.com/ethereum-optimism/specs/pull/93, we're in the process of defining what a standard chain means. 

This pull request is meant to introduce the terms in his PR but not be prescriptive about what values they should be. 

**Tests**

N/A

**Additional context**

Configurability Changes based on Ben's Notion document:  https://www.notion.so/Superchain-Segmentation-33a3152ca1dd4e87a40af52e08aae75d 
Relevant slack thread: https://oplabs-pbc.slack.com/archives/C06KERUPJ8H/p1710872632376849 

**Metadata**

N/A

### CODE REVIEW

File: specs/protocol/configurability.md
Comment (protolambda) [inline specs/protocol/configurability.md]: This is not configurable, but rather a metric of performance of the batcher efficiency. Can you split out metrics in its own table? Or remove this property?

File: specs/protocol/configurability.md
Comment (protolambda) [inline specs/protocol/configurability.md]: "Superchain target" | Choice of cross-L2 configuration. May be omitted in isolated OP Stack deployments. Includes SuperchainConfig and ProtocolVersions contract addresses."

Or something to that extent should probably be part of the table.

File: specs/protocol/configurability.md
Comment (protolambda) [inline specs/protocol/configurability.md]:
```suggestion
| Sequencing window                     | Maximum allowed batch submission gap, after which L1 fallback is triggered in derivation.                                                                                                                                                            |
```

High frequency implies many batch submissions. The fallback kicks in when it's the opposite, i.e. when there aren't any batch submissions for an extended amount of time.

File: specs/glossary.md
Comment (tynes) [inline specs/glossary.md]: Can we resolve this todo before merge?

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: We should differentiate between consensus and policy here. Consensus is enforced while policy includes the set of possible behaviors within consensus. An example of policy is the tx p2p gossip size limit, where consensus will accept a tx of any size given it fits in the block gas limit

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: This is policy

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: This is policy

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: The contract addresses themselves are consensus but the bytecode at the 1967 implementations are policy

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: This is better defined as the block at which the system config was initialized the first time

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: This is not a predeploy and therefore not present on all chains

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md:58]: What do you mean by "contracts" here? Do you mean contracts owned by L2 Proxy admin? ie all predeploys

File: specs/glossary.md
Comment (protolambda) [inline specs/glossary.md]:
```suggestion
## Batch submission frequency

Within the [sequencing-window] constraints the batcher is free to submit L2 blocks to for data-availability at any time. The batcher may be reconfigured with a new policy at any time by its operator. The rollup enforces safety guarantees and liveness through the sequencing window, if the batcher does not submit data within this allotted time.

By submitting new L2 data in smaller more frequent steps, there is less delay in confirmation of the L2 block inputs. This allows verifiers to ensure safety of L2 blocks sooner. This also reduces the time to finality of the data on L1, and thus the time to L2 input-finality.
 
By submitting new L2 data in larger less frequent steps, there is more time to aggregate more L2 data, and thus reduce fixed overhead of the batch-submission work. This can reduce batch-submission costs, especially for lower throughput chains that do not fill data-transactions (typically 128 KB of calldata, or 800 KB of blobdata) as quickly.
```

File: specs/protocol/configurability.md
Comment (protolambda) [reply specs/protocol/configurability.md]: Also needs a reference to the glossary entry.

File: specs/glossary.md
Comment (protolambda) [reply specs/glossary.md]: Needs TOC update.

File: specs/protocol/configurability.md
Comment (blmalone) [inline specs/protocol/configurability.md:21]: @tynes With this definition, it sounds some of the line items in the `Consensus Parameters` table below are out of place now. 

Anything that can be upgraded with admin keys should fall under policy right?

File: specs/protocol/configurability.md
Comment (blmalone) [inline specs/protocol/configurability.md]: @tynes new table for policy params

File: specs/protocol/configurability.md
Comment (blmalone) [inline specs/protocol/configurability.md]: @tynes is this more inline with how you were thinking?

File: specs/protocol/configurability.md
Comment (blmalone) [reply specs/protocol/configurability.md:21]: The structure of Ben's doc is: https://www.notion.so/Superchain-Segmentation-33a3152ca1dd4e87a40af52e08aae75d

Screenshot 2024-03-20 at 17 42 47

Comment (tynes) [reply specs/protocol/configurability.md:21]: > Anything that can be upgraded with admin keys should fall under policy right?

Not necessarily - consensus enforces the particular batcher address via the "batcher hash" config. The identity that owns the system config isn't related to consensus, it can be a smart contract or eoa altho we recommend a safe smart contract wallet

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: Perhaps remove the superset sentence because policy is constrained by consensus, meaning that its a subset. One example of a block builder policy would be "include only 1 tx per sender per block" where consensus allows an arbitrary number of txs per sender per block

File: specs/protocol/configurability.md
Comment (tynes) [inline specs/protocol/configurability.md]: The output frequency is policy. Making it smaller means less time after doing a withdrawal on L2 until you can prove your withdrawal on L1

File: specs/protocol/configurability.md
Comment (tynes) [reply specs/protocol/configurability.md]: It is important to clarify these as the things that can be changed by anybody and don't impact standardness

File: specs/protocol/configurability.md
Comment (blmalone) [reply specs/protocol/configurability.md]: The table structure has changed slightly based on Mark's comments. Can you clarify what you mean with the above comment?

Comment (protolambda) [reply specs/protocol/configurability.md]: I am suggesting to add an entry to the table

Comment (blmalone) [reply specs/protocol/configurability.md]: Added superchain target into the consensus parameters table: https://github.com/ethereum-optimism/specs/pull/98/files#diff-f4cc226d6bd3074bf41cc708de428f532feff29c5c7f901b9cc5a5864f8a68b5R50

File: specs/glossary.md
Comment (ben-chain) [inline specs/glossary.md]:
```suggestion
Within the [sequencing-window] constraints the batcher is permitted by the protocol to submit L2 blocks to for data-availability at any time. The batcher software allows for dynamic policy configuration by its operator. The rollup enforces safety guarantees and liveness through the sequencing window, if the batcher does not submit data within this allotted time.
```

Comment (ben-chain) [reply specs/glossary.md]: Bit of a nit, but I think this is a little more neutral/objective way to say it. This could be restricted to only "safe" values (e.g. Slipper's suggestions in notion) in the std-config; so "free to" might be a misnomer in that context.

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]:
```suggestion
- **Policy Parameters**: Parameters of the chain that might be changed without breaking consensus. Consensus is enforced by the protocol, while policy parameters may be changeable within constraints imposed by consensus.
```

Comment (ben-chain) [reply specs/protocol/configurability.md]: Same kinda note as other language suggeston.  I think the original wording is right most of the time, but strictly speaking, it is possible to impose social restrictions on acceptable policy that are more strict than consensus/protocol.

**Review by protolambda — APPROVED** [review]

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]: I meant to comment on notion, but I think this entry is too broad/vague.
- I believe there are a set of contracts which are NOT changeable or configurable based on the version. These should not be in the configurability table. For example, changing the L1 bridge code should not be thought of as "configuring" the standard version, it should be thought of creating a new version.
- That above set could be most or all of the L1 contracts, I'm actually not sure if there's anything that does not fulfill the above. Unless @tynes or @protolambda can think of any important contracts which don't meet the above definition, I would propose just removing this entry from the table.

**Review by ben-chain — CHANGES_REQUESTED** [review]
@blmalone this is a great start, thanks for putting together!!  Left some wording comments and a proposal to strike one entry pending others' review.

There is some information not ported over from Notion which I would like to see.  Specifically, can you add the `Administrator` column from the table?  Not opinionated on the name of the column, but for all of these config values, it is important to know who/what the protocol allows to change the config, if any.

Personally, I really liked having the concrete link to code as well, but I think that's a nice-to-have.

File: specs/glossary.md
Comment (mds1) [inline specs/glossary.md]: nit, extra word
```suggestion
Within the [sequencing-window] constraints the batcher is permitted by the protocol to submit L2 blocks for
data-availability at any time. The batcher software allows for dynamic policy configuration by its operator.
```

File: specs/protocol/configurability.md
Comment (mds1) [inline specs/protocol/configurability.md]: @ben-chain Why are fee vault contracts explicitly called out, but not other L2 predeploys? (this was carried over from your standard config strawman)

File: specs/protocol/configurability.md
Comment (mds1) [reply specs/protocol/configurability.md]: I think we do still want to include L1 contracts here because there is a corresponding requirement that to be a standard chain you must be using a governance approved version of the L1 contracts.

In other words, the implementations are technically a configurable part of the OP Stack, but any change means you are not a standard chain, and we need to convey that requirement. If we remove this entry from the table we can't convey that.

File: specs/protocol/configurability.md
Comment (ben-chain) [reply specs/protocol/configurability.md]: @mds1 good callout. @blmalone This can be removed in favor of overall predeploys entry, and then we restrict in standard config.

File: specs/protocol/configurability.md
Comment (ben-chain) [reply specs/protocol/configurability.md]: @mds1 I follow what you're saying, but I still feel pretty strongly that this is (or should be) a *versioning* difference, not a configuration difference.

I think the right way to convey the requirement is by requiring a version, which implies smart contracts, not by requiring certain bounds on config, which includes contracts.  As you said, "you must be using a governance approved _version_ of the L1 contracts".

It is a little less explicit in this PR's context, but I think it's also a valuable norm to say that changes to L1 contracts are a version difference, not a config difference.

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]:
```suggestion
  changeable through a privileged account or protocol upgrade.
```

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]:
```suggestion
```
Can we delete this entry? The DA layer is not configurable currently other than call vs blobdata, which I'll leave another comment on

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md:48]: Can we add an entry here for wherever the sequencer can choose between blobdata and calldata?

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]:
```suggestion
```
Let's remove this as well, not configurable away from ETH in current protocol version

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]:
```suggestion
```

Per @mds1 's comment this should be removed

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]:
```suggestion
```
Let's remove this -- that gaslimit setting is somewhat complex and in practice the target gas rate is downstream of this value AND sequencer behavior being set.  Base is actively working on this.

Since this row is part of the ResourceConfig, I think it's fine to just have an entry for resourceconfig.

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md:59]: Noting that my requested changes above would need to be reflected in removals here. It would be really nice to have a script to populate the `administers` field, but nice to have

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]: Duplicative with challenge period, can remove
```suggestion
```

File: specs/protocol/configurability.md
Comment (ben-chain) [inline specs/protocol/configurability.md]: can we specify that the batcher role is the administrator of this and the compression ratio

**Review by ben-chain — CHANGES_REQUESTED** [review]
Shaping up -- few more requested changes (apologies, should've caught some of these sooner)

File: specs/protocol/configurability.md
Comment (blmalone) [reply specs/protocol/configurability.md]: https://github.com/ethereum-optimism/specs/pull/98/commits/74eb66775dc76013b8288886219dad0185c1f626

File: specs/protocol/configurability.md
Comment (blmalone) [reply specs/protocol/configurability.md:59]: Sounds really useful, do we have an existing pattern for using scripts with `.md` files in the specs repo?

File: specs/protocol/configurability.md
Comment (blmalone) [reply specs/protocol/configurability.md:48]: https://github.com/ethereum-optimism/specs/pull/98/commits/e2c0da2376b2870f9b182a985d0c7ce64e58e52f

I'm basing this off the docs: https://docs.optimism.io/builders/chain-operators/management/blobs#update-your-batcher-to-post-calldata 

Specifically the batchers env var: `OP_BATCHER_DATA_AVAILABILITY_TYPE`

File: specs/protocol/configurability.md
Comment (ben-chain) [reply specs/protocol/configurability.md:59]: no idea xD.  Doesn't need to block this PR.

**Review by ben-chain — APPROVED** [review]
Woohoo!

**Pull request merged**
